File: crates/backforge-core/src/project/schema.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Тип поля таблицы
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum FieldType {
    Text,
    Integer,
    BigInt,
    Float,
    Boolean,
    Uuid,
    Timestamp,
    Json,
    Bytes,
}

/// Одно поле таблицы
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Field {
    pub id: Uuid,
    pub name: String,
    pub field_type: FieldType,
    pub nullable: bool,
    pub unique: bool,
    pub primary_key: bool,
    pub default_value: Option<String>,
}

impl Field {
    pub fn new(name: impl Into<String>, field_type: FieldType) -> Self {
        Self {
            id: Uuid::new_v4(),
            name: name.into(),
            field_type,
            nullable: true,
            unique: false,
            primary_key: false,
            default_value: None,
        }
    }

    pub fn primary_key(mut self) -> Self {
        self.primary_key = true;
        self.nullable = false;
        self
    }

    pub fn not_null(mut self) -> Self {
        self.nullable = false;
        self
    }
}

/// Таблица в схеме
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Table {
    pub id: Uuid,
    pub name: String,
    pub fields: Vec<Field>,
    pub indexes: Vec<Index>,
}

impl Table {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            id: Uuid::new_v4(),
            name: name.into(),
            fields: Vec::new(),
            indexes: Vec::new(),
        }
    }

    pub fn add_field(&mut self, field: Field) -> &mut Self {
        self.fields.push(field);
        self
    }
}

/// Индекс таблицы
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Index {
    pub id: Uuid,
    pub name: String,
    pub fields: Vec<String>,
    pub unique: bool,
}

/// Вся схема проекта (все таблицы)
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct ProjectSchema {
    pub tables: Vec<Table>,
}

File: services/code-generator/internal/schema/schema.go
// Package schema defines the Go representation of BackForge's project_state.json.
// Must stay in sync with crates/backforge-core/src/project/schema.rs and state.rs.
package schema

// FieldType mirrors FieldType in Rust core.
type FieldType string

const (
	FieldTypeText      FieldType = "text"
	FieldTypeInteger   FieldType = "integer"
	FieldTypeBigInt    FieldType = "big_int"
	FieldTypeFloat     FieldType = "float"
	FieldTypeBoolean   FieldType = "boolean"
	FieldTypeUUID      FieldType = "uuid"
	FieldTypeTimestamp FieldType = "timestamp"
	FieldTypeJSON      FieldType = "json"
	FieldTypeBytes     FieldType = "bytes"
)

type Field struct {
	ID           string     `json:"id"`
	Name         string     `json:"name"`
	FieldType    FieldType  `json:"field_type"`
	Nullable     bool       `json:"nullable"`
	Unique       bool       `json:"unique"`
	PrimaryKey   bool       `json:"primary_key"`
	DefaultValue *string    `json:"default_value,omitempty"`
}

type Index struct {
	ID     string   `json:"id"`
	Name   string   `json:"name"`
	Fields []string `json:"fields"`
	Unique bool     `json:"unique"`
}

type Table struct {
	ID      string  `json:"id"`
	Name    string  `json:"name"`
	Fields  []Field `json:"fields"`
	Indexes []Index `json:"indexes"`
}

type ProjectSchema struct {
	Tables []Table `json:"tables"`
}

type ProjectMeta struct {
	ID          string `json:"id"`
	Name        string `json:"name"`
	Description string `json:"description"`
	Version     uint32 `json:"version"`
}

type ProjectState struct {
	Meta   ProjectMeta    `json:"meta"`
	Schema ProjectSchema  `json:"schema"`
	OAuth  *OAuthProviders `json:"oauth,omitempty"`
}

// OAuthProviders holds optional OAuth provider credentials to inject into generated code.
type OAuthProviders struct {
	GitHub *GitHubOAuthConfig `json:"github,omitempty"`
}

// GitHubOAuthConfig contains the non-secret GitHub OAuth App values stored per-project.
// The client_secret is intentionally excluded and must be supplied at runtime via env var.
type GitHubOAuthConfig struct {
	ClientID    string `json:"client_id"`
	CallbackURL string `json:"callback_url"`
}
